### PR TITLE
Use sccache and clang-tidy-cache with S3 cache backend

### PR DESCRIPTION
**Pull request by Felixoid**
### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Use sccache as a replacement for ccache and using S3 as cache backend

### CODE REVIEW

**Comment by Felixoid** [comment]
Dear colleagues, it looks pretty solid and can be reviewed.

The cache in S3 works. The benefits we'd get out of it would be very significant, since the cache will be shared across all the RPs, and it will allow for reducing build time.

**Comment by rschu1ze** [comment]
> I've tested it locally with env SCCACHE_S3_NO_CREDENTIALS=true and CTCACHE_S3_READONLY=true, and it works inside the container. So everybody can have access to the shared cache

Are you saying, devs can speed up local builds with the shared cache?

**Comment by rschu1ze** [comment]
Too sleepy, I'll check tomorrow.

**Comment by Felixoid** [comment]
Yes, try the command `CMAKE_FLAGS='-DENABLE_CLICKHOUSE_SELF_EXTRACTING=1' ./packager --output-dir=../../build --package-type=binary --compiler=clang-15 --build-type=debug --clang-tidy --cache=sccache --s3-bucket=clickhouse-builds --docker-image-version=46240-bbb64a1d08940b88a77d7e30829f2ad6eae4f7a0 --version=23.3.1.1` from `docker/packager` for the current HEAD

File: cmake/ccache.cmake [hunk 1]
```diff
@@ -8,45 +8,57 @@ if (CMAKE_CXX_COMPILER_LAUNCHER MATCHES "ccache" OR CMAKE_C_COMPILER_LAUNCHER MA
...
+    message(STATUS "Using *ccache: no (disabled via configuration)")
     return()
 endif()
 
-find_program (CCACHE_EXECUTABLE ccache)
+# We support ccache and sccache
+if (FORCE_SCCACHE)
+    find_program (CCACHE_EXECUTABLE sccache)
+else()
+    find_program (CCACHE_EXECUTABLE ccache sccache)
```
Comment (rschu1ze) [inline cmake/ccache.cmake:23]: I am confused, should there really be a fallback from ccache to sccache? The call will be through `CACHE_EXECUTABLE`, are both binaries drop-replacements (exchangeable)? Maybe the logic should be less implicit (see my above comment).

File: cmake/ccache.cmake [hunk 2]
```diff
@@ -8,45 +8,57 @@ if (CMAKE_CXX_COMPILER_LAUNCHER MATCHES "ccache" OR CMAKE_C_COMPILER_LAUNCHER MA
...
 
 if (NOT CCACHE_EXECUTABLE)
-    message(${RECONFIGURE_MESSAGE_LEVEL} "Using ccache: no (Could not find find ccache. To significantly reduce compile times for the 2nd, 3rd, etc. build, it is highly recommended to install ccache. To suppress this message, run cmake with -DENABLE_CCACHE=0)")
+    message(${RECONFIGURE_MESSAGE_LEVEL} "Using *ccache: no (Could not find find ccache or sccache. To significantly reduce compile times for the 2nd, 3rd, etc. build, it is highly recommended to install one of them. To suppress this message, run cmake with -DENABLE_CCACHE=0)")
     return()
 endif()
 
-execute_process(COMMAND ${CCACHE_EXECUTABLE} "-V" OUTPUT_VARIABLE CCACHE_VERSION)
-string(REGEX REPLACE "ccache version ([0-9\\.]+).*" "\\1" CCACHE_VERSION ${CCACHE_VERSION})
+if (CCACHE_EXECUTABLE MATCHES "/ccache$")
```
Comment (rschu1ze) [inline cmake/ccache.cmake:32]: Yeah, that's special handling for ccache and sccache (l. 60). You might as well introduce variables `CCACHE_EXECUTABLE` and `SCCACHE_EXECUTABLE` then.

File: docker/packager/binary/Dockerfile [hunk 3]
```diff
@@ -97,7 +98,7 @@ ENV PATH="$PATH:/usr/local/go/bin"
 ENV GOPATH=/workdir/go
 ENV GOCACHE=/workdir/
 
-ARG CLANG_TIDY_SHA1=03644275e794b0587849bfc2ec6123d5ae0bdb1c
+ARG CLANG_TIDY_SHA1=c191254ea00d47ade11d7170ef82fe038c213774
```
Comment (rschu1ze) [inline docker/packager/binary/Dockerfile:101]: This is the version SHA of the clang-tidy-cache, right? Wondering why the existing one doesn't work with sccache?

File: docker/packager/packager [hunk 4]
```diff
@@ -361,17 +385,34 @@ if __name__ == "__main__":
...
     )
-    parser.add_argument("--distcc-hosts", nargs="+")
+    parser.add_argument(
+        "--s3-bucket",
+        help="a bucket with cache for clang-tidy-cache and sscache",
+    )
+    parser.add_argument(
+        "--s3-directory",
+        default="ccache",
+        help="a directory prefix with cache for clang-tidy-cache and sscache",
```
Comment (rschu1ze) [inline docker/packager/packager:407]: "a S3 directory prefix used for sscache and clang-tidy-cche"?

File: docker/test/fasttest/Dockerfile [hunk 5]
```diff
@@ -20,12 +20,6 @@ RUN apt-get update \
         zstd \
     --yes --no-install-recommends
 
-# Install CMake 3.20+ for Rust compilation
-RUN apt purge cmake --yes
```
Comment (rschu1ze) [inline docker/test/fasttest/Dockerfile:24]: Hm, why is cmake no longer build into the fastest dockerfile?

File: docker/test/fasttest/run.sh [hunk 6]
```diff
@@ -171,14 +179,14 @@ function run_cmake
     export CCACHE_COMPILERCHECK=content
     export CCACHE_MAXSIZE=15G
 
-    ccache --show-stats ||:
+    ccache_status
```
Comment (rschu1ze) [inline docker/test/fasttest/run.sh:182]: Could `ccache --zero-stats` (next line) also be moved to function `ccache-status`? (which should by the way now be called `compiler_cache_status`

File: tests/queries/0_stateless/01293_show_clusters.reference [hunk 7]
```diff
@@ -1,2 +1,3 @@
 test_shard_localhost
-test_shard_localhost	1	1	1	localhost	::1	9000	1	default	
+test_cluster_one_shard_two_replicas	1	1	1	127.0.0.1	127.0.0.1	9000	1	default
```
Comment (rschu1ze) [inline tests/queries/0_stateless/01293_show_clusters.reference:2]: unrelated?

**Comment by rschu1ze** [comment]
> Yes, try the command `CMAKE_FLAGS='-DENABLE_CLICKHOUSE_SELF_EXTRACTING=1' ./packager --output-dir=../../build --package-type=binary --compiler=clang-15 --build-type=debug --clang-tidy --cache=sccache --s3-bucket=clickhouse-builds --docker-image-version=46240-bbb64a1d08940b88a77d7e30829f2ad6eae4f7a0 --version=23.3.1.1` from `docker/packager` for the current HEAD

Worked for me, and there was also some caching (I cancelled the build and the next start picked up where it left).

However, few devs use `packager` locally. Can sccache also be used with plain CMake \+ ninja?

File: tests/queries/0_stateless/01293_show_clusters.reference (hunk 7 above)
Comment (Felixoid) [reply tests/queries/0_stateless/01293_show_clusters.reference:2]: Very related to `--network=host` in fasttest python script

File: docker/test/fasttest/run.sh (hunk 6 above)
Comment (Felixoid) [reply docker/test/fasttest/run.sh:182]: No, `--zero-stats` should be executed only once, it's intended.

File: docker/test/fasttest/Dockerfile (hunk 5 above)
Comment (Felixoid) [reply docker/test/fasttest/Dockerfile:24]: It's in the source image `test-utils`

File: docker/packager/binary/Dockerfile (hunk 3 above)
Comment (Felixoid) [reply docker/packager/binary/Dockerfile:101]: It's for https://github.com/matus-chochlik/ctcache/issues/17

File: docker/packager/packager [hunk 8]
```diff
@@ -387,34 +428,54 @@ if __name__ == "__main__":
...
         logging.info("Should place %s to output", args.with_binaries)
 
-    dockerfile = os.path.join(ch_root, "docker/packager", IMAGE_TYPE, "Dockerfile")
-    image_with_version = image_name + ":" + args.docker_image_version
-    if not check_image_exists_locally(image_name) or args.force_build_image:
-        if not pull_image(image_with_version) or args.force_build_image:
-            build_image(image_with_version, dockerfile)
+    if args.cache == "sccache":
+        if not args.s3_bucket:
+            raise argparse.ArgumentTypeError("sccache must have --s3-bucket set")
```
[the remark this answers is not in the document]
Comment (Felixoid) [reply docker/packager/packager:448]: Don't get it. The `--s3-directory` is optional, and could be empty as well

File: cmake/ccache.cmake (hunk 2 above)
Comment (Felixoid) [reply cmake/ccache.cmake:32]: Only one should be used in the end, so I don't see how implementing a few parameters would help.

**Comment by Felixoid** [comment]
> Can sccache also be used with plain CMake + ninja?
One must set a very specific set of ENV variables:

```
SCCACHE_S3_KEY_PREFIX=ccache/sccache
SCCACHE_S3_NO_CREDENTIALS=true
SCCACHE_BUCKET=clickhouse-builds
```

File: cmake/ccache.cmake (hunk 1 above)
Comment (rschu1ze) [reply cmake/ccache.cmake:23]: Well, we are not operating a nuclear power plant here 😄  The only effect would be unexpectedly slow builds. And CMake will print a warning if an option (`-D...`) was passed but not used, so users will notice.

Anyways, a cleaner separation of ccache  and sccache would be nice but it's up to you.

**Comment by rschu1ze** [comment]
> One must set a very specific set of ENV variables:
> 
> ```
> SCCACHE_S3_KEY_PREFIX=ccache/sccache
> SCCACHE_S3_NO_CREDENTIALS=true
> SCCACHE_BUCKET=clickhouse-builds
> ```

Maybe it is worth to document the steps [here](https://clickhouse.com/docs/en/development/developer-instruction) or [here](https://clickhouse.com/docs/en/development/build)? (depending of whether devs *should* use sscache or if sscache is only used for infra builds)

**Review by rschu1ze — APPROVED** [review]

**Comment by Felixoid** [comment]
Using it locally won't help at all, actually. The path must match. So, it's useful only with `packager`, as far as I understand the `*ccache` architect

File: docker/packager/packager [hunk 9]
```diff
@@ -243,32 +256,46 @@ def parse_env_variables(
...
-    if distcc_hosts:
-        hosts_with_params = [f"{host}/24,lzo" for host in distcc_hosts] + [
-            "localhost/`nproc`"
-        ]
-        result.append('DISTCC_HOSTS="' + " ".join(hosts_with_params) + '"')
-    elif cache == "distcc":
-        result.append('DISTCC_HOSTS="localhost/`nproc`"')
+        if s3_bucket:
+            # see https://github.com/matus-chochlik/ctcache#environment-variables
+            ctcache_dir = "clang-tidy-cache"
```
[the remark this answers is not in the document]
Comment (Felixoid) [reply docker/packager/packager:292]: It's a variable w/ very-very narrow scope. I'd rather make it `ctc_dir` to shorten it even more `¯\_(ツ)_/¯`

`ctcache` is the name of the product, so let's keep it as it is

File: cmake/ccache.cmake [hunk 10]
```diff
@@ -8,45 +8,57 @@ if (CMAKE_CXX_COMPILER_LAUNCHER MATCHES "ccache" OR CMAKE_C_COMPILER_LAUNCHER MA
     return()
 endif()
 
-option(ENABLE_CCACHE "Speedup re-compilations using ccache (external tool)" ON)
+option(ENABLE_CCACHE "Speedup re-compilations using ccache (default) or sccache (external tools)" ON)
```
[the remark this answers is not in the document]
Comment (Felixoid) [reply cmake/ccache.cmake]: It's addressed in https://github.com/ClickHouse/ClickHouse/pull/46240/commits/8efa2ce84390eaa26e8c718660273de08032e98a

File: cmake/ccache.cmake (hunk 1 above)
Comment (Felixoid) [reply cmake/ccache.cmake]: It's addressed in https://github.com/ClickHouse/ClickHouse/commit/8efa2ce84390eaa26e8c718660273de08032e98a

File: cmake/ccache.cmake (hunk 10 above)
[the remark this answers is not in the document]
Comment (rschu1ze) [reply cmake/ccache.cmake]: Merci.

File: cmake/ccache.cmake [hunk 11]
```diff
@@ -8,45 +9,62 @@ if (CMAKE_CXX_COMPILER_LAUNCHER MATCHES "ccache" OR CMAKE_C_COMPILER_LAUNCHER MA
     return()
 endif()
 
-option(ENABLE_CCACHE "Speedup re-compilations using ccache (external tool)" ON)
+option(ENABLE_CCACHE "Deprecated, use COMPILER_CACHE=(auto|ccache|sccache|disabled)")
```
Comment (rschu1ze) [inline cmake/ccache.cmake:12]: Maybe I don't see it, but where is `ENABLE_CCACHE` evaluated (besides l. 13-15)?

Comment (Felixoid) [reply cmake/ccache.cmake:12]: `OFF` by default for options https://cmake.org/cmake/help/latest/command/option.html

Comment (rschu1ze) [reply cmake/ccache.cmake:12]: What I meant is that `ENABLE_CCACHE` is not evaluated at all in this file. Whatever the user sets or doesn't set, it won't have an effect. This is not "deprecating" an option, it is abolishing it.

Comment (Felixoid) [reply cmake/ccache.cmake:12]: It's bool and can have either 1 or 0. How could we deprecate it without setting it to 0 and looking for 1?

If it's 1|ON, it's equal to default `COMPILER_CACHE=auto`. If it's unset, well, one will replace it in their script

**Comment by Felixoid** [comment]
Ok, I hope, it's the last rebasing
